### PR TITLE
(PUP-2929) Optimize function loader performance

### PR DESCRIPTION
**Pull request by thallgren**
This PR contains a series of performance improvements. Each commit covers one specific improvement. The details of each improvement can be found in the commit message.

The 'Use setXXX instead of repeated addXXX ...' commit includes some changes to RGen. A PR with those changes has been submitted here: https://github.com/mthiede/rgen/pull/22

### CODE REVIEW

File: lib/puppet/pops/loader/loader.rb [hunk 1]
```diff
@@ -137,43 +137,51 @@ def initialize(typed_name, value, origin)
...
+        @name = name
       end
+      @name_parts = parts
 
-      freeze()
-    end
+      # Not allowed to have numeric names - 0, 010, 0x10, 1.2 etc.
+      # At this point it is known that @name is a relativized string so it's
+      # safe to do a direct regexp match
+      unless Puppet::Pops::Patterns::NUMERIC.match(@name).nil?
```
Comment (hlindberg) [inline lib/puppet/pops/loader/loader.rb]: I wonder if the loader/typed-name really needs to care about numeric names - I don't think there is any ambiguity in the loaders. Names of things should be checked at the source (i.e. "illegal function name", "illegal class name" etc.) as there are far fewer of those than instances of TypedName.

That was a long way of saying: I think we can simply skip this assertion.

(sorry, I first commented on the commit - does not show up in the conversation).

File: lib/puppet/pops/adaptable.rb [hunk 2]
```diff
@@ -184,7 +187,11 @@ def self.associate_adapter(adapter, o)
     # @private
     #
     def self.instance_var_name(name)
-      name.split("::").join('_')
+      name.split(DOUBLE_COLON).join(USCORE)
+    end
+
+    def self.self_attr_name
```
Comment (hlindberg) [inline lib/puppet/pops/adaptable.rb]: Add documentation and mark it as api private

File: lib/puppet/vendor/rgen/lib/rgen/metamodel_builder/builder_extensions.rb [hunk 3]
```diff
@@ -395,7 +395,7 @@ def get<%= firstToUpper(name) %>
```
Comment (hlindberg) [inline lib/puppet/vendor/rgen/lib/rgen/metamodel_builder/builder_extensions.rb]: Never modify the vendored version directly! Users can install other versions of a vendored library - they are not forced to use what we ship. Patching has to be done through normal coding or worst case monkey patching just as if the library was not vendored.

Comment (hlindberg) [inline lib/puppet/vendor/rgen/lib/rgen/metamodel_builder/builder_extensions.rb]: We have some other patches / additions to rgen - see parser_factory.rb. There is also the vendored code loading that could be used to patch the vendored version IIRC.

File: lib/puppet/pops/visitor.rb [hunk 4]
```diff
@@ -24,11 +24,14 @@ def initialize(receiver, message, min_args=0, max_args=nil)
 
   # Visit the configured receiver
   def visit(thing, *args)
-    visit_this(@receiver, thing, *args)
+    visit_this(@receiver, thing, args)
   end
 
+  DOUBLE_COLON = '::'
+  NO_ARGS = []
```
Comment (hlindberg) [inline lib/puppet/pops/visitor.rb]: should be frozen since it is passed on in calls, and faulty code could add values to it (which would be very bad).

File: lib/puppet/pops/types/types.rb [hunk 5]
```diff
@@ -23,18 +23,7 @@ class PAnyType < TypeModelObject
       module ClassModule
         # Produce a deep copy of the type
         def copy
-          return @clone unless @clone.nil?
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: Ideally this revert should go into its own commit

**Comment by hlindberg** [comment]
Are there any benchmark results that show how much this speeds things up?

**Comment by hlindberg** [comment]
Are benchmarks added to the repo yet? They should go in first to enable before/after comparisson.

**Comment by hlindberg** [comment]
Found the relevant benchmarks. Looks like a 4% improvement on type inference and a whopping 3.6x improvement on function loading.

**Comment by thallgren** [comment]
Benchmarks have been added. I mention that in the JIRA.

**Comment by thallgren** [comment]
The 4% improvement on type inference is actually a lot more (about 50%) since the previous code was based on the flawed deep copy. Try without the optimization and with the marshal/unmarshal version of the copy. That will give you the real numbers.

File: lib/puppet/vendor/rgen_patch.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
+# Patch to improve generated addXXX and setXXX method performance
+# Submitted to rgen as PR: https://github.com/mthiede/rgen/pull/22
+#
+# The current implementation of the generated addXXX method will iterate
+# over the contained array where each iteration will then test for object_id
+# equality. Once the iteration is done and such an entry is found it
+# checks it the added value is an instance of MMBase or MMGeneric. If it
```
Comment (hlindberg) [inline lib/puppet/vendor/rgen_patch.rb]: typo "it"... "it", the second should be an "if" I think.

File: lib/puppet/vendor/rgen_patch.rb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+#
+# This patch changes the implementation so that the test for MMBase/
+# MMGeneric is made prior to the iteration, thus avoiding the iteration
+# alltogether for all non model objects
+#
+# The patch also changes the setXXX method to allow it to optimize the way
+# it ensures that the added elements are unique. Calling the addXXX method
+# to do an increasingly expensive sequential scan is very expensive when
+# large sets of values are assigned.
+#
```
Comment (hlindberg) [inline lib/puppet/vendor/rgen_patch.rb]: worth mentioning and reference the PR to RGen and that this patch can then be removed.

**Comment by hlindberg** [comment]
Nice, this has an 11% speedup on mixed evaluation as well!

File: lib/puppet/vendor/rgen_patch.rb (hunk 7 above)
Comment (thallgren) [inline lib/puppet/vendor/rgen_patch.rb]: It is mentioned. Second line from the top.
